lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,582 @@
+		// -- disable propagate_variable_equivalences in tail_simplifier
+		params.set (":xform.tail_simplifier_pve", false);
+		params.set (":xform.subsumption_checker", true);
+//		params.set (":order_children", true);
+//		params.set (":pdr.max_num_contexts", "500");
+		fp.set (params);
+		new_db.loadZFixedPoint (fp, false);
+		boost::tribool result = fp.query ();
+
+		LOG("pabs-smt2", outs() << "SMT2: " << fp << "\n";);
+
+		if (result) outs () << "sat";
+		else if (!result)
+		{
+			outs() << "unsat\n";
+			HornDbModel absModel;
+			initDBModelFromFP(absModel, new_db, fp);
+
+			converter.convert(absModel, oldModel);
+			LOG("pabs-debug", outs() << "FINAL RESULT:\n";);
+			//Print invariants
+			printInvars(M, oldModel);
+		}
+		else outs () << "unknown";
+		outs () << "\n";
+		Stats::stop("Pabs solve");
+
+		return false;
+	}
+
+	void PredicateAbstraction::getAnalysisUsage (AnalysisUsage &AU) const
+	{
+		AU.addRequired<HornifyModule> ();
+		AU.setPreservesAll();
+	}
+
+	void PredicateAbstraction::printInvars (Module &M, HornDbModel &origModel)
+	{
+		 for (auto &F : M) printInvars (F, origModel);
+	}
+
+	void PredicateAbstraction::printInvars(Function &F, HornDbModel &origModel)
+	{
+		  if (F.isDeclaration ()) return;
+
+		  HornifyModule &hm = getAnalysis<HornifyModule> ();
+		  outs () << "Function: " << F.getName () << "\n";
+
+		  // -- not used for now
+		  Expr summary = hm.summaryPredicate (F);
+
+		  ZFixedPoint<EZ3> fp = *m_fp;
+
+		  for (auto &BB : F)
+		  {
+			if (!hm.hasBbPredicate (BB)) continue;
+
+			Expr bbPred = hm.bbPredicate (BB);
+
+			outs () << *bind::fname (bbPred) << ":";
+			const ExprVector &live = hm.live (BB);
+			//Expr invars = fp.getCoverDelta (bind::fapp (bbPred, live));
+			Expr invars = origModel.getDef(bind::fapp(bbPred, live));
+
+			if (isOpX<AND> (invars))
+			{
+			  outs () << "\n\t";
+			  for (size_t i = 0; i < invars->arity (); ++i)
+				outs () << "\t" << *invars->arg (i) << "\n";
+			}
+			else
+			  outs () << " " << *invars << "\n";
+		  }
+	}
+
+	HornClauseDB PredicateAbstractionAnalysis::generateAbstractDB(HornClauseDB &db, PredAbsHornModelConverter &converter)
+	{
+		HornClauseDB new_DB(db.getExprFactory ());
+
+		generateAbstractRelations(db, new_DB, converter);
+
+		generateAbstractRules(db, new_DB, converter);
+
+		generateAbstractQueries(db, new_DB);
+
+		LOG("pabs-debug", outs() << "NEW DB: \n";);
+		LOG("pabs-debug", outs() << new_DB << "\n";);
+
+		converter.setAbsDB(new_DB); //set converter
+
+		return new_DB;
+	}
+
+	void PredicateAbstractionAnalysis::generateAbstractRelations(HornClauseDB &db, HornClauseDB &new_DB, PredAbsHornModelConverter &converter)
+	{
+		//For each relation, generate its abstract version
+		for(Expr rel : db.getRelations())
+		{
+			LOG("pabs-debug", outs() << "OLD REL: " << *rel << "\n";);
+			ExprVector new_args;
+			//Push fdecl name
+			Expr old_fdecl_name = bind::fname(rel);
+			std::ostringstream oss;
+			oss << old_fdecl_name << "_pabs";
```

Comment:
do not convert expressions to strings. You probably wanted to use `tag` expressions here.